app/src/main/java/io/github/trojan_gfw/igniter/tile/IgniterTileService.java

```diff
@@ -0,0 +1,159 @@
+    }
+
+    @Override
+    public void onTestResult(String testUrl, boolean connected, long delay, @NonNull String error) {
+        // Do nothing, since TileService will not submit test request.
+    }
+
+    @Override
+    public void onBinderDied() {
+        LogHelper.i(TAG, "onBinderDied");
+    }
+
+    private void updateTile(final @ProxyService.ProxyState int state) {
+        Tile tile = getQsTile();
+        if (tile == null) {
+            return;
+        }
+        LogHelper.i(TAG, "updateTile with state: " + state);
+        switch (state) {
+            case ProxyService.STATE_NONE:
+                tile.setState(Tile.STATE_INACTIVE);
+                tile.setLabel(getString(R.string.app_name));
+                break;
+            case ProxyService.STARTED:
+                tile.setState(Tile.STATE_ACTIVE);
+                tile.setLabel(getString(R.string.tile_on));
+                break;
+            case ProxyService.STARTING:
+                tile.setState(Tile.STATE_ACTIVE);
+                tile.setLabel(getString(R.string.tile_starting));
+                break;
+            case ProxyService.STOPPED:
+                tile.setState(Tile.STATE_INACTIVE);
+                tile.setLabel(getString(R.string.tile_off));
+                break;
+            case ProxyService.STOPPING:
+                tile.setState(Tile.STATE_INACTIVE);
+                tile.setLabel(getString(R.string.tile_stopping));
+                break;
+            default:
+                break;
+        }
+        tile.updateTile();
+    }
+
+    @Override
+    public void onClick() {
+        super.onClick();
+        LogHelper.i(TAG, "onClick");
+        if (MultiProcessSP.isFirstStart(true)) {
+            // if user never open Igniter before, when he/she clicks the tile, it is necessary
+            // to start the launcher activity for resource preparation.
+            Intent intent = new Intent(this, MainActivity.class);
+            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+            startActivity(intent);
+            return;
+        }
+        ITrojanService service = mConnection.getService();
+        if (service == null) {
+            mTapPending = true;
+        } else {
+            try {
+                @ProxyService.ProxyState int state = service.getState();
+                switch (state) {
+                    case ProxyService.STARTED: {
+                        startActivity(ProxyControlActivity.startOrStopProxy(this, false, false));
+                        break;
+                    }
+                    case ProxyService.STARTING:
+                        break;
+                    case ProxyService.STATE_NONE:
+                    case ProxyService.STOPPED: {
+                        startActivity(ProxyControlActivity.startOrStopProxy(this, true, false));
+                        break;
+                    }
+                    case ProxyService.STOPPING:
+                        break;
```

Comment:
default case and throw exceptions here

Reply:
Throwing exception in TileService may cause bad experience. Assume that the default case occurs, each time user slides down the menu for tile, system will show a dialog stating that Igniter is crashed again. I suggest logging here.

Reply:
LGTM.

Reply:
Added log already.